sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java

```diff
@@ -490,6 +530,56 @@ private HttpPipeline setupPipeline(HttpClient httpClient) {
             });
     }
 
+    /**
+     * Gets token from shared token cache
+     * */
+    public Mono<AccessToken> authenticateWithSharedTokenCache(TokenRequestContext request, String username) {
+        // find if the Public Client app with the requested username exists
+        return Mono.fromFuture(() -> getPublicClientApplication(true).getAccounts())
+                .onErrorMap(t -> new CredentialUnavailableException(
+                        "Cannot get accounts from token cache. Error: " + t.getMessage(), t))
+                .flatMap(set -> {
+                    IAccount requestedAccount;
+                    Map<String, IAccount> accounts = new HashMap<>(); // home account id -> account
+
+                    for (IAccount cached : set) {
+                        if (username == null || username.equals(cached.username())) {
+                            if (!accounts.containsKey(cached.homeAccountId())) { // only put the first one
+                                accounts.put(cached.homeAccountId(), cached);
+                            }
+                        }
+                    }
+
+                    if (accounts.size() == 0) {
+                        if (username == null) {
+                            return Mono.error(new CredentialUnavailableException("No accounts were discovered in the "
+                                    + "shared token cache. To fix, authenticate through tooling supporting azure "
+                                    + "developer sign on."));
+                        } else {
+                            return Mono.error(new CredentialUnavailableException(String.format("User account '%s' was "
+                                    + "not found in the shared token cache. Discovered Accounts: [ '%s' ]", username,
```

Comment:
Error messages and checks were changed here and below. 
This one is PII exposure.

Reply:
Thanks! I lost the update during the merge.